plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/tenantdata/TenantDataCacheLoader.java

```diff
@@ -0,0 +1,111 @@
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.ForbiddenException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.UnauthorizedException;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.rest.HttpJsonRequestFactory;
+import org.eclipse.che.api.core.rest.HttpJsonResponse;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+
+class TenantDataCacheLoader extends CacheLoader<TenantDataCacheKey, UserCheTenantData> {
+
+  private final boolean standalone;
+  private final String cheNamespace;
+  private final HttpJsonRequestFactory httpJsonRequestFactory;
+  private final String fabric8UserServiceEndpoint;
+
+  @Inject
+  TenantDataCacheLoader(
+      HttpJsonRequestFactory httpJsonRequestFactory,
+      @Named("che.fabric8.user_service.endpoint") String fabric8UserServiceEndpoint,
+      @Named("che.infra.openshift.project") String cheNamespace,
+      @Named("che.fabric8.standalone") boolean standalone) {
+    this.fabric8UserServiceEndpoint = fabric8UserServiceEndpoint;
+    this.cheNamespace = cheNamespace;
+    this.standalone = standalone;
+    this.httpJsonRequestFactory = httpJsonRequestFactory;
+  }
+
+  @Override
+  public UserCheTenantData load(TenantDataCacheKey cacheKey) throws InfrastructureException {
+    if (standalone) {
+      return loadStandalone(cacheKey);
+    }
+
+    String responseBody;
+    try {
+      responseBody = getResponseBody(fabric8UserServiceEndpoint, cacheKey.getKeycloakToken());
+    } catch (Exception e) {
+      throw new InfrastructureException("Exception during the user tenant data retrieval", e);
+    }
+    try {
+      final Gson gson = new Gson();
+      final UserServicesJsonResponse userServicesData =
+          gson.fromJson(responseBody, UserServicesJsonResponse.class);
+      List<Namespace> namespaces = userServicesData.getNamespaces();
+      for (Namespace ns : namespaces) {
+        if (cacheKey.getNamespaceType().equals(ns.getType())) {
+          UserCheTenantData cheTenantData =
+              new UserCheTenantData(
+                  ns.getName(),
+                  ns.getClusterUrl(),
+                  ns.getClusterAppDomain(),
+                  ns.isClusterCapacityExhausted());
+          UserCheTenantDataValidator.validate(cheTenantData);
+          return cheTenantData;
+        }
+      }
+    } catch (ValidationException | JsonSyntaxException e) {
+      throw new InfrastructureException(
+          "Invalid response from Fabric8 user services:" + responseBody, e);
+    }
+    throw new InfrastructureException(
+        format(
+            "No namespace with type '%s' was found in the user tenant",
+            cacheKey.getNamespaceType()));
+  }
+
+  private UserCheTenantData loadStandalone(TenantDataCacheKey cacheKey) {
+    String namespaceType = cacheKey.getNamespaceType();
```

Comment:
We could move this to the retriever since caching is not needed for this case.

Reply:
You're right, that's cleaner.